provider/equinix/environ.go

```diff
@@ -78,31 +81,100 @@ func (e *environ) Config() *config.Config {
 }
 
 func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
-	return nil, errors.NewNotImplemented(nil, "not implemented")
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
 }
 
 func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
-	return nil, errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-is-controller": "true", "juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return nil, err
+	}
+	instanceIDs := make([]instance.Id, len(insts))
+	for i, inst := range insts {
+		instanceIDs[i] = inst.Id()
+	}
+	return instanceIDs, nil
 }
 
 func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
 	return nil
 }
 
 func (e *environ) Destroy(ctx context.ProviderCallContext) error {
-	return errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, inst := range insts {
+		if _, err = e.equinixClient.Devices.Delete(string(inst.Id()), true); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return common.Destroy(e, ctx)
 }
 
 func (e *environ) DestroyController(ctx context.ProviderCallContext, controllerUUID string) error {
-	return errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return err
+	}
+
+	for _, inst := range insts {
```

Comment:
Also, the logic between `Destroy` and `DestroyController` is pretty much the same with the difference being the instance selection query. I would recommend extracting the common bits into a `e.destroyInstances([]instance.Id) error` method and just have the aforementioned methods select the appropriate instances and pass them to the helper.

Reply:
Yes, it is a for loop but you can find a copy of it in at least three methods. If we extract it, we can potentially handle errors in a more consistent fashion.

Reply:
Done thanks